### ISSUE TITLE
import-order should allow wrapper functions to be declared earlier.

### ISSUE THREAD

**Issue by jamestalmage** [benmosher/eslint-plugin-import#383]
Example using `pify` (promisifies nodeback style API's):

```js
var pify = require('pify'); // eslint-disable-line import-order/import-order
var fs = pify(require('fs'));
```

It would be nice if it was smart enough to figure out this should be allowed.

**Comment by benmosher** [comment]
Would configuring the rule reordering core modules after externals solve this?

Seems like in a project using pify, you would be doing that generally anyway?

**Comment by jamestalmage** [comment]
> Would configuring the rule reordering core modules after externals solve this?

It might make the error go away, but I use the default ordering across many projects. That consistency is worth more to me than having to write an occasional eslint-disable comment. But it would be nice if the plugin was just smart enough to handle it for me.

**Comment by benmosher** [comment]
> _just_ smart enough to handle it for me

(emphasis mine. 😉)

What smart behavior would you expect? The simplest things I can think of would be to

- ignore `require` calls that are part of a function call's arguments. thus,
  - `require('pify')` must still appear in `externals` section
  - `pify(require('fs')` may appear anywhere thereafter

or

- detect that a module is a context-based type, ("module modifier"?): if it is a function that has `require` calls as arguments, it can be (_must_ be?) `import`/`require`'d before any other imports/requires. thus,
  - `pify(require('fs'))` must still appear in `builtins` as `require('fs')` would otherwise be
  - `require('pify')` can/must appear before any other requires/imports

The latter seems ideal to me. Also straightforward to discern; just need to wait to report any out-of-place upfront `require`s until it is clear they are not called with other `require`s.

**Comment by jamestalmage** [comment]
I think either solution works, but I think the second one is better.

>  it can be (must be?) import/require'd before any other imports/requires

I would vote `must be`. That makes it explicit and prevents an unnecessary diff later on if you decide to you want to `pify`-ify something higher in the list.

**Comment by sindresorhus** [comment]
I wouldn't do this. With `import` syntax it won't be possible anymore. Better to just adjust and make it easy to convert the code to `import` later on.

You should rather do:

```js
var pify = require('pify');
var fs = require('fs');
var fsP = pify(fs);
```

This also has the benefit of making it clear in the code that the `fs` is a promisified version.

**Comment by jamestalmage** [comment]
Yeah that's probably better. I like the `P` convention
